File: samples/pulse.cpp
#include <cmath>
#include <esn/esn.hpp>
#include <iomanip>
#include <iostream>
#include <random>

static const unsigned kNeuronCount = 100;
static const float kStep = 0.01f;
static const float kIntervalWidthMin = 0.1f;
static const float kIntervalWidthMax = 1.0f;
static const float kTargetIntervalWidth = 0.7f;
static const float kTargetIntervalError = 0.1f;
static const float kOutputPulseWidth = 0.1f;
static const float kOutputPulseMax = 1.0f;
static const float kOutputPulseThreshold = 0.0001f;
static const unsigned kTrainPulseCount = 100;

inline float Pulse( float x, float width, float max )
{
    float retval = max * std::exp( -std::pow(
        6.0f * ( x - width / 2.0f ) / width, 2.0f ) );
    return ( retval < kOutputPulseThreshold ) ? 0.0f : retval;
}

int main()
{
    ESN::NetworkParamsNSLI params;
    params.neuronCount = kNeuronCount;
    params.inputCount = 1;
    params.outputCount = 1;
    auto network = ESN::CreateNetwork( params );

    bool inputState = false;
    std::vector< float > input( 1 );
    float currentIntervalWidth = 0.0f;
    float nextInterval = 0.0f;
    std::default_random_engine randomEngine;
    std::uniform_real_distribution< float > random(
        kIntervalWidthMin, kIntervalWidthMax );

    std::vector< float > referenceOutput( 1 );
    float outputPulseStart = 0.0f;
    float outputPulseAmplitude = 0.0f;
    unsigned outputPulseCount = 0;

    std::vector< float > output( 1 );

    for ( float time = 0.0f; true; time += kStep )
    {
        if ( time > nextInterval )
        {
            if ( inputState &&
                 std::fabs( currentIntervalWidth - kTargetIntervalWidth ) <
                 kTargetIntervalError )
            {
                outputPulseStart = time;
                outputPulseAmplitude = ( kTargetIntervalError - std::fabs(
                    currentIntervalWidth - kTargetIntervalWidth ) ) /
                    kTargetIntervalError * kOutputPulseMax;
                outputPulseCount ++;
            }

            inputState = !inputState;
            input[0] = inputState ? 1.0f : 0.0f;

            currentIntervalWidth = random( randomEngine );
            nextInterval += currentIntervalWidth;
        }

        referenceOutput[0] = Pulse( time - outputPulseStart,
            kOutputPulseWidth, outputPulseAmplitude );

        network->SetInputs( input );
        network->Step( kStep );
        network->CaptureOutput( output );

        if ( outputPulseCount < kTrainPulseCount )
            network->TrainOnline( referenceOutput, true );

        std::cout <<
            std::setw( 15 ) << time <<
            std::setw( 3 ) << inputState <<
            std::setw( 15 ) << referenceOutput[0] <<
            std::setw( 15 ) << output[0] <<
            std::endl;
    }

    return 0;
}

File: samples/pulse.py
from math import *
import imp
import random

esn = imp.load_source( "esn", "@ESN_PYTHON_MODULE@" )
esn.load_library( "@ESN_DLL@" )

# Check if we can use matplotlib
try :
    imp.find_module( "matplotlib" )
    from matplotlib import pyplot
    from matplotlib import animation
    USE_MATPLOTLIB = True
except ImportError :
    USE_MATPLOTLIB = False

NEURON_COUNT = 100
LEAKING_RATE = 0.1
CONNECTIVITY = 0.1
SIM_STEP = 0.01
INTERVAL_WIDTH_MIN = 0.1
INTERVAL_WIDTH_MAX = 1.0
TARGET_INTERVAL_WIDTH = 0.7
TARGET_INTERVAL_ERROR = 0.1
INPUT_PULSE_MAX = 1.0
OUTPUT_PULSE_WIDTH = 0.1
OUTPUT_PULSE_MAX = 1.0
OUTPUT_PULSE_THRESHOLD = 0.0001
TRAIN_PULSE_COUNT = 100
STEPS_PER_FRAME = 20

class Model :

    def __init__( self ) :

        self.network = esn.CreateNetworkNSLI(
            inputCount = 1,
            neuronCount = NEURON_COUNT,
            outputCount  = 1,
            leakingRate = LEAKING_RATE,
            connectivity = CONNECTIVITY
        )

        self.inputState = False
        self.inputs = [ 0 ]
        self.currentIntervalWidth = 0
        self.nextInterval = 0
        self.outputPulseStart = 0
        self.outputPulseAmplitude = 0
        self.outputPulseCount = 0

    def pulse( self, x, width, amplitude ) :

        retval = amplitude * exp( -pow( 6 * ( x - width / 2 ) / width, 2 ) )
        return 0 if retval < OUTPUT_PULSE_THRESHOLD else retval

    def simulate( self, frame ) :

        time = frame * SIM_STEP

        if time > self.nextInterval :
            error = abs( self.currentIntervalWidth - TARGET_INTERVAL_WIDTH )
            if self.inputState and error < TARGET_INTERVAL_ERROR :
                self.outputPulseStart = time
                self.outputPulseAmplitude = ( TARGET_INTERVAL_ERROR -   \
                    abs( self.currentIntervalWidth -                    \
                        TARGET_INTERVAL_WIDTH ) ) /                     \
                    TARGET_INTERVAL_ERROR * OUTPUT_PULSE_MAX
                self.outputPulseCount += 1

            self.inputState = not self.inputState
            self.inputs[0] = 1 if self.inputState else 0

            self.currentIntervalWidth = random.uniform( INTERVAL_WIDTH_MIN,
                INTERVAL_WIDTH_MAX )
            self.nextInterval += self.currentIntervalWidth

        self.inputs[0] = self.pulse( self.nextInterval - time,
            self.currentIntervalWidth, INPUT_PULSE_MAX ) if self.inputState else 0

        referenceOutput = self.pulse( time - self.outputPulseStart,
            OUTPUT_PULSE_WIDTH, self.outputPulseAmplitude );

        self.network.set_inputs( self.inputs )
        self.network.step( SIM_STEP )
        output = self.network.capture_output( 1 )

        if self.outputPulseCount < TRAIN_PULSE_COUNT :
            self.network.train_online( [ referenceOutput ], True );

        print( "%10s %7s %10s %10s" %
                (
                    str( "%0.5f" % time ),
                    str( "%r" % self.inputState ),
                    str( "%0.5f" % referenceOutput ),
                    str( "%0.5f" % output[0] )
                )
            )

        return time, self.inputs[0], referenceOutput, output[0]

model = Model()

if USE_MATPLOTLIB :

    figure = pyplot.figure()
    subplot = figure.add_subplot( 111 )
    inputLine, refLine, outputLine = subplot.plot( [], [], [], [], [], [] )
    subplot.set_ylim( -0.1, 1.1 )
    subplot.grid( True )

    timeData = []
    inputData = []
    refData = []
    outputData = []

    def animationFunc( frame ) :
        for i in range( 0, STEPS_PER_FRAME ) :
            time, input, ref, output = \
                model.simulate( i + STEPS_PER_FRAME * frame )
            timeData.append( time )
            inputData.append( input )
            refData.append( ref )
            outputData.append( output )
        inputLine.set_data( timeData, inputData )
        refLine.set_data( timeData, refData )
        outputLine.set_data( timeData, outputData )
        subplot.set_xlim( time - 1, time + 0.1 )

    anim = animation.FuncAnimation( figure, animationFunc, interval = 30 )

    pyplot.show()

else :

    frame = 0
    while True :
        frame += 1
        model.simulate( frame )
